File: backend-go/internal/platform/bus/bus.go
package bus

import "sync"

// In-process sync event bus (modular monolith integration events).
type Bus struct {
	mu       sync.RWMutex
	handlers map[string][]func(any) error
}

func New() *Bus {
	return &Bus{handlers: map[string][]func(any) error{}}
}

func (b *Bus) Subscribe(eventType string, h func(any) error) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.handlers[eventType] = append(b.handlers[eventType], h)
}

func (b *Bus) Publish(eventType string, event any) error {
	b.mu.RLock()
	hs := append([]func(any) error{}, b.handlers[eventType]...)
	b.mu.RUnlock()
	for _, h := range hs {
		if err := h(event); err != nil {
			return err
		}
	}
	return nil
}

File: backend-python/app/architecture/bus.py
"""In-process sync event bus (modular monolith integration events)."""

from __future__ import annotations

from collections import defaultdict
from collections.abc import Callable
from typing import Any


Handler = Callable[[Any], None]


class Bus:
    def __init__(self) -> None:
        self._handlers: dict[str, list[Handler]] = defaultdict(list)

    def subscribe(self, event_type: str, handler: Handler) -> None:
        self._handlers[event_type].append(handler)

    def publish(self, event_type: str, event: Any) -> None:
        for handler in list(self._handlers.get(event_type, [])):
            handler(event)


bus = Bus()
